packages/terminal/src/browser/terminal-quick-open-service.ts

```diff
@@ -0,0 +1,155 @@
+import { TerminalWidget } from './base/terminal-widget';
+import { TerminalService } from './base/terminal-service';
+import { TerminalCommands } from './terminal-frontend-contribution';
+
+@injectable()
+export class TerminalQuickOpenService implements QuickOpenModel, QuickOpenHandler {
+
+    @inject(PrefixQuickOpenService)
+    protected readonly prefixQuickOpenService: PrefixQuickOpenService;
+
+    @inject(CommandRegistry)
+    protected readonly commandRegistry: CommandRegistry;
+
+    @inject(TerminalService)
+    protected readonly terminalService: TerminalService;
+
+    readonly prefix: string = 'term ';
+
+    get description(): string {
+        return 'Show All Opened Terminals';
+    }
+
+    getModel(): QuickOpenModel {
+        return this;
+    }
+
+    getOptions(): QuickOpenOptions {
+        return {
+            fuzzyMatchLabel: {
+                enableSeparateSubstringMatching: true
+            },
+            fuzzyMatchDescription: {
+                enableSeparateSubstringMatching: true
+            }
+        };
+    }
+
+    open(): void {
+        this.prefixQuickOpenService.open(this.prefix);
+    }
+
+    async onType(lookFor: string, acceptor: (items: QuickOpenItem[]) => void): Promise<void> {
+        const terminalItems: QuickOpenItem[] = [];
+
+        // Get the list of currently opened terminal widgets, sorted by the widget id.
+        const widgets: TerminalWidget[] = this.terminalService.all
+            .sort((a: TerminalWidget, b: TerminalWidget) => (a.id.localeCompare(b.id)));
+
+        for (const widget of widgets) {
+            const item = await this.toItem(widget);
+            terminalItems.push(item);
+        }
+        // Append a quick open item to create a new terminal.
+        const createNewTerminalItem = new QuickOpenItem({
+            label: 'Open New Terminal',
+            run: this.doCreateNewTerminal()
+        });
+        terminalItems.push(createNewTerminalItem);
+
+        acceptor(terminalItems);
+        return;
+    }
+
+    /**
+     * Get the function that can create a new terminal.
+     * @param {TerminalWidget} widget - the terminal widget to be opened.
+     * @returns Function that would create a new terminal if mode === QuickOpenMode.OPEN.
+     */
+    protected doCreateNewTerminal(): (mode: QuickOpenMode) => boolean {
+        return (mode: QuickOpenMode) => {
+            if (mode !== QuickOpenMode.OPEN) {
+                return false;
+            }
+            this.commandRegistry.executeCommand(TerminalCommands.NEW.id);
+            return true;
+        };
+    }
+
+    /**
+     * Convert the terminal widget to the quick open item.
+     * @param {TerminalWidget} widget - the terminal widget.
+     * @returns The quick open item.
+     */
+    protected async toItem(widget: TerminalWidget): Promise<QuickOpenItem<QuickOpenItemOptions>> {
+
+        const options: QuickOpenItemOptions = {
+            label: widget.id,
```

Comment:
Are you sure we should display the `id` as the quick-open label? The terminal's `id` contains much more information, so I was just wondering what your thoughts were.

Currently, when running a task, we get the following items in the `term ` quick-open.

<div align='center'>

<img width="602" alt="Screen Shot 2019-06-26 at 8 46 31 AM" src="https://user-images.githubusercontent.com/40359487/60180968-2aa5d600-97ef-11e9-9e06-33f0169b923a.png">


</div>

Reply:
I was unsure about what to be displayed. I thought the `title label` (`gitpod/workspace/...` may be too verbose as the label, so I decided to display the `widget id` (terminal-#) as the label and the `title label` as the description.

Reply:
For comparison, here's vscode:

<div align='center'>

<img width="520" alt="Screen Shot 2019-06-26 at 9 55 29 AM" src="https://user-images.githubusercontent.com/40359487/60185808-93458080-97f8-11e9-9b4f-5803004ae35b.png">

</div>

I guess in our case I don't see the added value of displaying the `terminal-0` as the first label.
Maybe it's because in our case our `id` is actually `terminal-0` and not simply `0`.

Reply:
@lmcbout do you have an opinion on this?